core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -4976,14 +4993,49 @@ private static void computeVersion(ServletContext context) {
      * Version number of this Jenkins.
      */
     public static String VERSION = UNCOMPUTED_VERSION;
-
+    
+    /**
+     * Minimal Java level required to run this Jenkins.
+     * If {@code null}, the version is unknown.
+     */
+    private static @CheckForNull String JAVA_MIN_LEVEL = null;
+    private static VersionNumber DEFAULT_JAVA_MIN_LEVEL = new VersionNumber("8");
+    
+    /**
+     * Allows setting custom minimal Java level.
+     * If {@code null}, Jenkins will try to determine the version automatically.
+     * @since TODO
+     */
+    @VisibleForTesting
+    /*package*/ static final @Nonnull String ENFORCED_MIN_JAVA_LEVEL_PROPERTY_NAME = 
+            Jenkins.class.getName() + ".javaMinLevel";
+    
     /**
      * Parses {@link #VERSION} into {@link VersionNumber}, or null if it's not parseable as a version number
      * (such as when Jenkins is run with "mvn hudson-dev:run")
      */
     public @CheckForNull static VersionNumber getVersion() {
         return toVersion(VERSION);
     }
+    
+    /**
+     * Gets Java minimal level required to run this Jenkins instance.
+     * @return Minimal Java level or {@code null} if it cannot be determined
+     * @since TODO
+     */
+    public @CheckForNull static VersionNumber getJavaMinLevel() {
+        return toVersion(JAVA_MIN_LEVEL);
```

Comment:
And if a plugin is installed that uses Java9?  if you expect this to then return Java9 then what if that plugin only executes code on the master and never on an agent?  Why would that then force callers (the downstream ssh PR) to use a higher agent version? unnesceserily.  At this point I am not sure what this is trying to acheive - and no property is needed as this can be found out at runtime as you have all the information you need in the class file.

Reply:
> And if a plugin is installed that uses Java9? if you expect this to then return Java9 then what if that plugin only executes code on the master and never on an agent

We consider Jenkins to be a shared context platform. If the class is not annotated as `@OnMaster`, it may be actually sent to the agent by another class. It's hard to control that. As a happy Remoting maintainer, I would be against supporting different Java versions on master and agents.

We could add that (just via 2 properties), but IMHO it needs a real use-case. And, even there is such case, the support can be added later in a compatible way.